src/server/stream_family.cc

```diff
@@ -1402,44 +1533,131 @@ std::optional<ReadOpts> ParseReadArgsOrReply(CmdArgList args, ConnectionContext*
     return std::nullopt;
   }
 
+  vector<unordered_map<string_view, StreamIDsItem>> res_sitems(shard_set->size());
+  bool stop = false;
+  auto g_cb = [&](Transaction* t, EngineShard* shard) {
+    if (!opts.read_group) {
+      return OpStatus::OK;
+    }
+    auto sid = shard->shard_id();
+    auto s_args = t->GetShardArgs(sid);
+    unordered_map<string_view, StreamIDsItem> sid_items;
+
+    // get group and consumer
+    for (size_t i = 0; i < s_args.size(); i++) {
+      string_view key = s_args[i];
+
+      streamCG* group = nullptr;
+      streamConsumer* consumer = nullptr;
+      auto group_res = FindGroup(t->GetOpArgs(shard), key, opts.gname);
+      if (!group_res) {
+        (*cntx)->SendError(group_res.status());
+        stop = true;
+        return group_res.status();
+      }
+      if (group = group_res->second; !group) {
+        (*cntx)->SendError(NoGroupError(key, opts.gname));
+        stop = true;
+        return OpStatus::SKIPPED;
+      }
+
+      shard->tmp_str1 = sdscpylen(shard->tmp_str1, opts.cname.data(), opts.cname.size());
+      consumer = streamLookupConsumer(group, shard->tmp_str1, SLC_NO_REFRESH);
+      if (!consumer) {
+        consumer =
+            streamCreateConsumer(group, shard->tmp_str1, NULL, 0, SCC_NO_NOTIFY | SCC_NO_DIRTIFY);
+      }
+      StreamIDsItem sitem;
+      sitem.group = group;
+      sitem.consumer = consumer;
+      auto [_, is_inserted] = sid_items.emplace(key, sitem);
+      if (!is_inserted) {
+        (*cntx)->SendError("Same stream specified multiple times");
+        stop = true;
+        return OpStatus::SKIPPED;
+      }
+    }
+    res_sitems[sid] = sid_items;
+    return OpStatus::OK;
+  };
+
+  cntx->transaction->Schedule();
+  cntx->transaction->Execute(std::move(g_cb), false);
+
+  if (stop) {
+    goto close_transaction;
+  }
+
+  for (auto& sitems : res_sitems) {
+    opts.stream_ids.merge(sitems);
+  }
+
   // Parse the stream IDs.
   for (size_t i = opts.streams_arg + streams_count; i < args.size(); i++) {
     string_view key = ArgS(args, i - streams_count);
     string_view idstr = ArgS(args, i);
 
+    StreamIDsItem sitem;
     ParsedStreamId id;
 
     if (idstr == "$") {
       // Set ID to 0 so if the ID cannot be resolved (when the stream doesn't
       // exist) it takes the first entry added.
+      if (opts.read_group) {
```

Comment:
Do you have to parse AFTER you start running a transaction?
It's much better to parse everything before you start with the transaction and return errors earlier. This way you know that you continue with consistent arguments. Also If you do this you could factor out this code into a helper function that will return failure/success and this way you won't need "goto".

``` 
if (!ParseArgs(args, cntx)) {
  return;
}
```

Reply:
Yeah, I was able to get rid of `goto` in my latest push. It is cleaner. Please take a look at it.